### ISSUE TITLE
Migrate inline String evaluators to OpenJ9

### ISSUE THREAD

**Issue by fjeremic** [eclipse/omr#2293]
As part of https://github.com/eclipse/omr/pull/2292 we had a change to fix negation of an `if`/`else` path. This change is related to String inline evaluators very specific to Java and should be migrated to OpenJ9 as they are not used and do not belong in OMR.

**Comment by MohamedBouzid** [comment]
I am interested to help solve the issue.

**Comment by 0xdaryl** [comment]
Thanks for your interest @MohamedBouzid !

There are a number of evaluator support functions defined in `compiler/z/codegen/OMRTreeEvaluator.cpp`.  They only have relevance to the OpenJ9 project and are only referenced from there:
```
TR::Register *inlineStringHashCode(TR::Node* node, TR::CodeGenerator* cg, bool isCompressed)
TR::Register *inlineStringHashCodeUnrolled(TR::Node* node, TR::CodeGenerator* cg)
TR::Register *inlineUTF16BEEncodeSIMD(TR::Node *node, TR::CodeGenerator *cg)
TR::Register *inlineUTF16BEEncode(TR::Node *node, TR::CodeGenerator *cg)
```
@fjeremic, please correct if I've missed any.  In fact, `inlineStringHashCodeUnrolled` doesn't appear to be used even in OpenJ9.  Is there value in keeping it?

All of these functions should be relocated to the file `runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp` in the OpenJ9 project where they are actually consumed.

Your plan of attack should look something like this:

1.  If you're not familiar with Eclipse OMR (this repo) or Eclipse OpenJ9 (https://github.com/eclipse/openj9) I suggest you first read the Contributor Guidelines found in the `CONTRIBUTING.md` file in the root directory of both projects.  Please follow the contribution rules and guidelines in your pull requests.

2.  In the Eclipse OpenJ9 project, create a pull request (PR) where those functions are relocated to `runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp`.  Change their definitions to be static functions and remove any `extern` declarations.  For example,
```
static TR::Register *inlineStringHashCode(TR::Node *node, TR::CodeGenerator *cg, bool isCompressed)
   {
   ...
   }
```
This is necessary because the linker will complain about duplicate symbols between the two projects otherwise.  Because this change is in the Z code generator, a friendly OpenJ9 committer will launch a Jenkins build targeting the Z platform to ensure it builds successfully.  The PR will be reviewed and merged.

3.  Once the PR from the previous step has been merged, create a PR in the Eclipse OMR project to delete the functions from `compiler/z/codegen/OMRTreeEvaluator.cpp`.  An OMR committer will launch a genie build targeting the Z platforms to ensure it builds successfully.  The PR will be reviewed and merged and this issue can be closed.  Note that you can create this PR before the Eclipse OMR PR is merged, but make sure you prefix the PR title with "WIP" to ensure an OMR committer does not merge this before you are ready.

I hope this helps.  Please reach out here if you have any troubles.  Thanks again for your participation!

**Comment by fjeremic** [comment]
> @fjeremic, please correct if I've missed any. In fact, inlineStringHashCodeUnrolled doesn't appear to be used even in OpenJ9. Is there value in keeping it?

I don't think so. I implemented this on zEC12 before we had vector instructions in z/Architecture as a textbook example of how to do Duff's Device in the the codegen with labels and branches. There is a comment detailing the outcome. We can remove it in a commit of it's own so it is easily searchable in the future. This can be done as part of this issue.
